cli/stats.go

```diff
@@ -0,0 +1,138 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/spf13/cobra"
+)
+
+const (
+	containerHeader     = "CONTAINER ID"
+	containerNameHeader = "NAME"
+	cpuPercHeader       = "CPU %"
+	netIOHeader         = "NET I/O"
+	blockIOHeader       = "BLOCK I/O"
+	memPercHeader       = "MEM %"
+	memUseHeader        = "MEM USAGE / LIMIT"
+	pidsHeader          = "PIDS"
+)
+
+// statsDescription is used to describe stats command in detail and auto generate command doc.
+var statsDescription = "stats command is to display a live stream of container(s) resource usage statistics"
+
+// StatsCommand use to implement 'stats' command
+type StatsCommand struct {
+	baseCommand
+
+	noStream bool
+	//TODO: add more flags support
+}
+
+// Init initialize stats command.
+func (stats *StatsCommand) Init(c *Cli) {
+	stats.cli = c
+	stats.cmd = &cobra.Command{
+		Use:   "stats [OPTIONS] [CONTAINER...]",
+		Short: "Display a live stream of container(s) resource usage statistics",
+		Long:  statsDescription,
+		Args:  cobra.MinimumNArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return stats.runStats(args)
+		},
+		Example: statsExample(),
+	}
+	stats.addFlags()
+}
+
+// addFlags adds flags for specific command.
+func (stats *StatsCommand) addFlags() {
+	flagSet := stats.cmd.Flags()
+	flagSet.BoolVar(&stats.noStream, "no-stream", false, "Disable streaming stats and only pull the first result")
+}
+
+// runStats is the entry of stats command.
+func (stats *StatsCommand) runStats(args []string) error {
+	ctx := context.Background()
+	apiClient := stats.cli.Client()
+	containers := args
+
+	cStats := []*StatsEntryWithLock{}
+	waitFirst := &sync.WaitGroup{}
+	for _, name := range containers {
+		s := &StatsEntryWithLock{StatsEntry: StatsEntry{container: name}}
+		cStats = append(cStats, s)
+		waitFirst.Add(1)
+		go collect(ctx, s, apiClient, !stats.noStream, waitFirst)
+	}
+
+	// before print to screen, make sure each container get at least one valid stat data
+	waitFirst.Wait()
+
+	// do a quick scan in case container not found and so on
+	var errs []string
+	for _, c := range cStats {
+		err := c.GetError()
+		if err != nil {
+			errs = append(errs, err.Error())
```

Comment:
What about tagged each error with their container id? like `$cid: $err`. User may be confused about which container the error belongs to .

Reply:
done, I wrapper the cid info in the `getError` func, please check it.